vnext/Microsoft.ReactNative/Utils/TextHitTestUtils.cpp

```diff
@@ -0,0 +1,176 @@
+    // not on the same line, we use the rightmost boundary of the TextBlock.
+    //
+    // The side-effect is that RTL text may have additional valid hit box
+    // space at the end of a line, but it's better than the alternative of the
+    // hit box excluding the last character on a line.
+    const auto prevPointer = textPointer.GetPositionAtOffset(-1, winrt::LogicalDirection::Forward);
+    if (prevPointer != nullptr) {
+      const auto prevRect = prevPointer.GetCharacterRect(winrt::LogicalDirection::Forward);
+      if (rect.Y == prevRect.Y) {
+        right = prevRect.X;
+      }
+    }
+  }
+
+  // The character is after the point if the Y-coordinate of the point is above
+  // (less than) the top of the character rect, or if the Y-coordinate is above
+  // (less than) the bottom of the character rect and the X-coordinate is
+  // "before" the beginning of the character rect (less than the left side for
+  // LTR or greater than the right for RTL):
+  // ┌────> X
+  // │ (x,y) ┌─────┐  ┌─────┐ (x,y)
+  // ↓ ┌─────┘     │  │     └─────┐
+  // Y └───────────┘  └───────────┘
+  //        LTR            RTL
+  const auto isBeforeX = !isRtl ? point.X < rect.X : point.X > right;
+  return point.Y < rect.Y || (point.Y < bottom && isBeforeX);
+}
+
+static bool IsRTL(const winrt::TextPointer &textPointer) {
+  auto currentPointer = textPointer.GetPositionAtOffset(1, winrt::LogicalDirection::Forward);
+  auto currentRect = textPointer.GetCharacterRect(winrt::LogicalDirection::Forward);
+  auto firstCharacterRect = currentRect;
+  while (currentPointer != nullptr) {
+    currentRect = currentPointer.GetCharacterRect(winrt::LogicalDirection::Forward);
+    // If we haven't figured out the direction by the time we reach the end of
+    // a line, try again for the next line.
+    if (currentRect.Y != firstCharacterRect.Y) {
+      firstCharacterRect = currentRect;
+    } else if (currentRect.X != firstCharacterRect.X) {
+      return currentRect.X < firstCharacterRect.X;
+    }
+
+    currentPointer = currentPointer.GetPositionAtOffset(1, winrt::LogicalDirection::Forward);
+  }
+
+  // Assume LTR if there are not enough characters to determine LTR vs. RTL.
+  // This may occur for TextBlocks with a single character per line.
+  return false;
+}
+
+static winrt::TextPointer GetPositionFromPointCore(
+    const winrt::TextPointer &start,
+    const winrt::TextPointer &end,
+    const winrt::Point &targetPoint) {
+  // Since characters in a TextBlock are sorted from top-left to bottom-right
+  // (or top-right to bottom-left for RTL), we can use binary search to find
+  // the character with bounds that container the pointer point.
+  //
+  // This algorithm currently makes the following assumptions:
+  // 1. Characters on the same line have the same Rect::Y value
+  // 2. Search space is over only LTR or only RTL characters
+  const auto width = start.VisualParent().Width();
```

Comment:
This text block width heuristic is problematic. I'm going to experiment with just using the character width again.

Reply:
Fixed. There is indeed a minor issue with characters at the edge of a word wrap boundary, but I don't believe we should worry about this for now.